Kubernetes Resources (#8308)
DataDog/documentation: content/en/infrastructure/livecontainers.md

If you're using Kubernetes, enable Kubernetes Resources for Live Containers to gain multi-dimensional visibility into all Kubernetes workloads across your clusters. Inspired by the `kubectl` tool, this feature gives you complete coverage of your Kubernetes infrastructure in a continuously updated table with curated resource metrics, faceted search, pre-workload detailed view, and visualized maps.

## Installation

Follow the [Docker][6] or [Kubernetes][7] Agent installation instructions. Container metrics are available without additional configuration after installation.

**Kubernetes Resources for Live Containers requires installation of**:

* [Datadog Agent][8] version 7.21.1 (or above)
* [Datadog Cluster Agent][9] 1.8.0 (or above)

### Kubernetes Resources

To enable Kubernetes Resources for Live Containers, follow the [Helm instructions][10] and add the following changes to your `values.yaml` file:

{{< code-block lang="yaml" >}}
datadog:
   ...
   orchestratorExplorer:
      enabled: true
   ...
   agents:
      image:
        repository: datadog/agent
        tag: 7.21.1
        pullPolicy: Always
   ...
   clusterAgent:
	repository: datadog/cluster-agent

> tag: 1.7.0

Reviewer: just to doublecheck. If this corresponds to the DCA version for orchestrator explorer, it should be `1.8.0`

Reviewer 2: actually both tags should be `tag: latest` now, and we specify the correct version in the requirements above (1.8.0 indeed)

Reviewer 2: we also have an example for the correct values in https://docs.google.com/document/d/1LiIvmynJ4hPsVytWSZXM3X3PZfXhVW9j0zcvVb3pi60/edit#

Reviewer 2: also note that lines 58-60 are not indented correctly

pullPolicy: Always
   ...
{{< /code-block >}}

## Configuration

### Include/Exclude containers

It is possible to include and/or exclude containers from real-time collection:

> * Exclude containers either via passing the environment variable `DD_CONTAINER_EXCLUDE` or adding `container_exclude:` in your `datadog.yaml` main configuration file.

Reviewer 3: ```suggestion
* Exclude containers either by passing the environment variable `DD_CONTAINER_EXCLUDE` or adding `container_exclude:` in your `datadog.yaml` main configuration file.
```

> * Include containers either via passing the environment variable `DD_CONTAINER_INCLUDE` or adding `container_include:` in your `datadog.yaml` main configuration file.

Reviewer 3: ```suggestion
* Include containers either by passing the environment variable `DD_CONTAINER_INCLUDE` or by adding `container_include:` in your `datadog.yaml` main configuration file.
```

      path: /opt/datadog-agent/run
    name: pointerdir

```shell
container_exclude: ["image:debian"]
container_include: ["name:frontend.*"]
```

**Note**: For Agent 5, instead of including the above in the `datadog.conf` main configuration file, explicitly add a `datadog.yaml` file to `/etc/datadog-agent/`, as the Process Agent requires all configuration options here. This configuration only excludes containers from real-time collection, **not** from Autodiscovery.

## Getting Started

Navigate to the [Containers page][1]. This will automatically bring you to the **Containers** view.

## Searching, Filtering, and Pivoting

### String Search

Containers are, by their nature, extremely high cardinality objects. Datadog's flexible string search matches substrings in the container name, ID, or image fields.

> If you've enabled Kubernetes Resources, strings such as Pod, deployment, ReplicaSet, and service name and Kubernetes labels are searchable in a [Kubernetes Resources view](#kubernetes-resources-views).

Reviewer 3: ```suggestion
If you've enabled Kubernetes Resources, strings such as `pod`, `deployment`, `ReplicaSet`, and `service name`, as well as Kubernetes labels are searchable in a [Kubernetes Resources view](#kubernetes-resources-views).
```

Use parentheses to group operators together. For example, `(NOT (elasticsearch OR kafka) java) OR python`.

### Filtering and Pivoting

> The screenshot below displays a system that has been filtered down to a Kubernetes cluster of 9 nodes. RSS and CPU utilization on containers is reported compared to the provisioned limits on the containers, when they exist. Here, it is apparent that the containers in this cluster are over-provisioned. You could use tighter limits and bin packing to achieve better utilization of resources.

Reviewer 3: ```suggestion
The screenshot below displays a system that has been filtered down to a Kubernetes cluster of nine nodes. RSS and CPU utilization on containers is reported compared to the provisioned limits on the containers, when they exist. Here, it is apparent that the containers in this cluster are over-provisioned. You could use tighter limits and bin packing to achieve better utilization of resources.
```

{{< img src="infrastructure/livecontainers/overprovisioned.png" alt="Over Provisioned"  style="width:80%;">}}

> Container environments are dynamic and can be hard to follow. The following screenshot displays a view that has been pivotted by `kube_service` and `host`—and, to reduce system noise, filtered to `kube_namespace:default`. You can see what services are running where, and how saturated key metrics are:

Reviewer 3: ```suggestion
Container environments are dynamic and can be hard to follow. The following screenshot displays a view that has been pivoted by `kube_service` and `host`—and, to reduce system noise, filtered to `kube_namespace:default`. You can see what services are running where, and how saturated key metrics are:
```

* [**Metrics**][15]: View live metrics for your resource. You can view any graph full screen, share a snapshot of it, or export it from this tab.
* **Processes**: View all processes running in containers of this resource.
* **Network**: View your resource’s network performance, including source, destination, sent and received volume, and throughput fields. Use the Destination field to search by tags like DNS or ip_type, or use the Group by filter in this view to group network data by tags, like pod_name or service.
* [**Traces**][16]: View traces for your resource, including the date, service, duration, method, and status code of a trace.
* [**Events**][17]: View all Kubernetes events for your resource.

For a detailed dashboard of this resource, click the View Pods Dashboard in the top right corner of this panel.

### Container Logs

View streaming logs for any container like `docker logs -f` or `kubectl logs -f` in Datadog. Click any container in the table to inspect it. Click the *Logs* tab to see real-time data from [Live Tail][18] or indexed logs for any time in the past.

#### Live Tail

> With Live Tail, all container logs are streamed -- pausing the stream allows you to easily read logs that are quickly being written; un-pause to continue streaming.

Reviewer 3: ```suggestion
With Live Tail, all container logs are streamed. Pausing the stream allows you to easily read logs that are quickly being written; unpause to continue streaming.
```

[2]: /integrations/docker_daemon/
[3]: /agent/kubernetes/
[4]: /integrations/amazon_ecs/
[5]: https://app.datadoghq.com/containers/kubernetes-beta
[6]: /agent/docker/#run-the-docker-agent
[7]: /agent/kubernetes/?tab=helm
[8]: /agent/
[9]: /agent/cluster_agent/setup/?tab=secret
[10]: /agent/kubernetes/?tab=helm#installation
[11]: /getting_started/tagging/
[12]: /getting_started/tagging/unified_service_tagging
[13]: /dashboards/widgets/timeseries/
[14]: /logs
[15]: /metrics
[16]: /tracing
[17]: /events
[18]: /logs/live_tail/
[19]: https://gist.github.com/hkaj/404385619e5908f16ea3134218648237

> [20]: https://github.com/DataDog/helm-charts

Reviewer: maybe use this link instead https://github.com/DataDog/helm-charts/tree/master/charts/datadog for a direct link to the configurable values
